### PR TITLE
Add IOF XML import & some CSV imports & some support for Eventor

### PR DESCRIPTION
**Pull request by arnost00**
Added to QE :
- IOF XML import
  - single race support (ORIS & IOF Eventor)
  - relays support (IOF Eventor)
  - imports: event (create new race), clubs, registration, classes, entries
- CSV imports (key is: Cze registration, runs.id or IOF ID), supported columns : SI, class, bib, start time
- CSV export for TV Graphics (single race)
- CSV result export extended
- EventConfig extend
  - flag for IOF Eventor race
  - race number for XML export
- IOF XML export update for Eventor
  - WMOC Startlist
  - WMOC Results
  - WC Startlist (single race, relays)
  - WC Results (single race, relays)

Fix relays results - relay summary run status.
Fix Txt Emma service - better time handling, added long club name to startist
Extend Txt Emma service - option for CSV starlist (uses [EmmaClient - release 2023-08-17](https://github.com/arnost00/liveresults/releases/tag/Release_2023-08-17))
OResults export - @otahirs - prevent export of invalid running time in case finish time is lower than start time and running time (timeMs) values are not valid
Card Reader plugin - RunFlagDialog is used instead of two checkboxes - fixing a possible inconsistency when checking/unchecking them

QE tested on IOF races : WC, WMOC, WMTBOC

Online service used : 
 * WC - OResults (BlueBox devices), 
 * WMOC - OResults (radio data from Racom uploaded with [racom_client, build 14)](https://github.com/oresults/racom_client) and EmmaClient (txt, uses [EmmaClient, release 2023-08-17](https://github.com/arnost00/liveresults/releases/tag/Release_2023-08-17))
 * WMTBOC - OResults (BlueBox devices), 

There is a quick document [how to work with QE & Eventor](https://github.com/arnost00/quickbox/wiki/qe_eventor) on my wiki page. After the merge, we can move it to the main wiki.

Custom print reports from WMOC [wmoc2023_custom_reports.zip](https://github.com/Quick-Event/quickbox/files/12458940/wmoc2023_custom_reports.zip)
Optimized for print on 1 column & with start numbers.

### CODE REVIEW

**Comment by fvacek** [comment]
I suggest to prepare it for qt6 branch.

**Comment by fvacek** [comment]
> > I suggest to prepare it for qt6 branch.
> 
> After switching the QE master branch to Qt6, I can switch that too. Now it seems better to develop in Qt5 (qmake)

I'm afraid, that you will have as many marge conflicts, that it will force you to stick to Qt5 version for the rest of your life, but it is your decision. :)

File: quickevent/app/quickevent/plugins/Oris/src/xmlimporter.h [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+protected:
+	bool readPersonNode(SPerson &s, QXmlStreamReader &reader, const XmlCreators creator);
+
+	bool importEntries(QXmlStreamReader &reader, const XmlCreators creator);
+	bool importStartlist(QXmlStreamReader &reader, const XmlCreators creator);
+	bool importClasses(QXmlStreamReader &reader, const XmlCreators creator);
+	bool importClubs(QXmlStreamReader &reader, const XmlCreators creator);
+	bool importRegistration(QXmlStreamReader &reader, const XmlCreators creator);
+
+	QString GenFakeCzClubAbbr(QString country);
```
Comment (fvacek) [inline quickevent/app/quickevent/plugins/Oris/src/xmlimporter.h:61]: `genFakeCzClubAbbr`

File: quickevent/app/quickevent/plugins/Oris/src/txtimporter.cpp [hunk 2]
```diff
@@ -238,3 +238,240 @@ void TxtImporter::importParsedCsv(const QList<QVariantList> &csv)
...
 	getPlugin<EventPlugin>()->emitDbEvent(Event::EventPlugin::DBEVENT_COMPETITOR_COUNTS_CHANGED);
 }
+
+int TxtImporter::getStartTimeMSec(QString str, int start00_msec)
+{
+	bool ok;
+	double dbl_time = str.toDouble(&ok);
+	int st_time = 0;
+	if(ok) {
+		st_time = ((int)dbl_time) * 60 + (((int)(dbl_time * 100)) % 100);
```
Comment (fvacek) [inline quickevent/app/quickevent/plugins/Oris/src/txtimporter.cpp:248]: use `static_cast`, `C` cast is deprecated by linters

File: quickevent/app/quickevent/plugins/Oris/src/xmlimporter.h [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+#include <QObject>
+#include <QXmlStreamReader>
+#include <QSet>
+#include <QMap>
+
+class XmlImporter : public QObject
+{
+	Q_OBJECT
+public:
+	enum class XmlCreators : int
```
Comment (fvacek) [inline quickevent/app/quickevent/plugins/Oris/src/xmlimporter.h:13]: `enum class XmlCreators` is enough

File: quickevent/app/quickevent/plugins/Oris/src/xmlimporter.h [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+
+		int siNumber = 0;
+
+		QString nationalityCode;	// Person
+		QString nationalityName;	// Person
+		QString clubCode;			// Organisation - Club
+		QString clubName;			// Organisation - Club
+		QString countryCode;		// Organisation - NationalFederation
+		QString countryName;		// Organisation - NationalFederation
+		int clubIdIof = -1;	// Eventor (Club or NationalFederation ID)
```
Comment (fvacek) [inline quickevent/app/quickevent/plugins/Oris/src/xmlimporter.h:40]: `std::optional<int> clubIdIof`

File: quickevent/app/quickevent/plugins/Oris/src/xmlimporter.h [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+		QSet<int> enterRaces;
+		// ...
+		int bib = 0;
+		int rank = 0;
+		int start_order_pref = 0; // 0-none /  1- early / 2- middle / 3- late
```
Comment (fvacek) [inline quickevent/app/quickevent/plugins/Oris/src/xmlimporter.h:45]: enum please

**Review by fvacek — CHANGES_REQUESTED** [review]

File: quickevent/app/quickevent/plugins/Oris/src/txtimporter.cpp (hunk 2 above)
Comment (fvacek) [reply quickevent/app/quickevent/plugins/Oris/src/txtimporter.cpp:248]: sure

File: quickevent/app/quickevent/plugins/Oris/src/xmlimporter.h (hunk 5 above)
Comment (arnost00) [reply quickevent/app/quickevent/plugins/Oris/src/xmlimporter.h:45]: Currently unused variables (rank, start_order_pref) - will be removed.

**Comment by fvacek** [comment]
@arnost00  still draft?

**Comment by arnost00** [comment]
> @arnost00 still draft?

Yes. Still are missing some parts (relay import & export / single race result export for eventor)
I'm working on it.

**Comment by otahirs** [comment]
ted po MS v MTBO bych z tohoto udelal posledni Qt5 release

**Comment by arnost00** [comment]
Fix receipt for free order - correct split times are shown. Not all controls in "Extra punches"
receipt_update

**Comment by fvacek** [comment]
> ted po MS v MTBO bych z tohoto udelal posledni Qt5 release

jeste by to chtelo zmergeovat :)

**Comment by arnost00** [comment]
Custom print reports for WMOC
[wmoc2023_custom_reports.zip](https://github.com/Quick-Event/quickbox/files/12458940/wmoc2023_custom_reports.zip)

Optimized for print on 1 column & with start numbers.

Print reports for :
 - Startlist for Starters
 - Startlist by classes
 - Results current stage

**Comment by otahirs** [comment]
any more changes coming? or what is needed before (merge and) release for the autumn season?

**Comment by fvacek** [comment]
Please merge this ASAP, this monster MR will be hard to port to QE3 :(

**Review by fvacek — APPROVED** [review]

**Comment by arnost00** [comment]
I will try to close it as soon as possible. And finish the missing things later.

**Comment by arnost00** [comment]
> any more changes coming? or what is needed before (merge and) release for the autumn season?

Only small fixes to CSV result export, and CSV import - CZE support fixed

**Comment by arnost00** [comment]
There are still a lot of things to do:
- Test @otahirs PR on relays XML import
- Test this PR on CZE race (prints / exports )
- Add missing support to XML import (CZE relays)
- missing TV graphics exports : relays (IOF, CZE), single race (CZE) 

I can move this things to separate tasks.

**Comment by otahirs** [comment]
> Test @otahirs [PR on relays XML import](https://github.com/arnost00/quickbox/pull/2)

this needs to be a separate task which involves rewriting bigger part of QE  
QE handle relay names as `club` + `number`, which cannot be used for all use cases.   
we should handle `name` separately and does not rely in exports on  `club` + `number` combination

**Comment by arnost00** [comment]
There is a quick document (how to work with QE & Eventor)[https://github.com/arnost00/quickbox/wiki/qe_eventor] on my wiki page.
After the merge, we can move it to the main wiki.

**Comment by fvacek** [comment]
In ideal word, any change should have one issue, one commit. If it is implemented in more commits because of review process, then all the commits should be squashed before merge, rebasing on master will be very appreciated.

This way the changes  can be relatively easily ported to `qe3`, what is desired state. Feature rich QE2 is not long term goal.

**Comment by arnost00** [comment]
I plan to make other new features for `QE3`.
I think that `QE2` will be used by users this fall and possibly winter.
And the development will be primarily for `QE3`.

**Comment by arnost00** [comment]
I'm starting to prepare wiki page about [How to organize IOF event using QuickEvent](https://github.com/Quick-Event/quickbox/wiki/How-to-organize-IOF-event-using-QuickEvent)

**Comment by otahirs** [comment]
@arnost00 should I do the rebasing into discrete commits and prepare new MR?

**Comment by arnost00** [comment]
> @arnost00 should I do the rebasing into discrete commits and prepare new MR?

Yes, you can, if you have time and energy.

**Comment by arnost00** [comment]
Fixed 2 bugs found on [MČR na krátké trati 2023](https://oris.orientacnisporty.cz/Zavod?id=6808)

**Comment by arnost00** [comment]
On the weekend of September 23 and 24, build 697 successfully used on ČPŠ (2 x relays - approx 400 teams) & ČP ŽA sprint.

**Comment by fvacek** [comment]
what about merge?

**Comment by arnost00** [comment]
> what about merge?
  
Yes. Merge it if possible

**Pull request merged**